bullet_train-fields/app/javascript/controllers/fields/file_field_controller.js

```diff
@@ -77,16 +82,61 @@ export default class extends Controller {
     this.removeFileFlagTarget.value = true;
   }
 
-  handleFileSelected() {
-    const statusText = this.selectFileButtonTarget.querySelector("span");
-    const icon = this.selectFileButtonTarget.querySelector("i");
-
+  handleFileSelected(event) {
     if (this.hasDownloadFileButtonTarget) {
       this.downloadFileButtonTarget.remove();
     }
 
-    statusText.innerText = "Select Another File";
-    icon.classList.remove("ti-upload");
-    icon.classList.add("ti-check");
+    const files = event.target.files;
+    this.updateSelectedFileList(files);
+  }
+
+  cancelFileUpload(event){
+    const fileInput = this.fileFieldTarget;
+    const files = fileInput.files;
+    const fileToCancel = event.target.dataset.filename;
+
+    const dt = new DataTransfer();
+    for (const file of files) {
+      if(file.name != fileToCancel){
+        dt.items.add(file);
+      }
+    }
+    fileInput.files = dt.files;
+    this.updateSelectedFileList(dt.files);
   }
+
+  updateSelectedFileList(files){
+    const statusText = this.selectFileButtonTarget.querySelector("span");
+    const icon = this.selectFileButtonTarget.querySelector("i");
+
+    this.selectedFileListTarget.innerHTML = "";
+
+    if(files.length){
+      for (const file of files) {
```

Comment:
The other approach is to package a sample row inside a `<template>` tag that is part of the html partial, that would be a target called `selectedFileRowTemplate`, with all of the classes and localized text elements defined in that `<template>` tag, and have text placeholders like `@FILENAME@` that you'd `innerHTML.replace` within the duplicated copy before inserting it in the right place in the DOM.

`<template` tags are by default hidden and can't be interacted with. I think that'd be super clean in this case, and would allow others to eject and update the styles of this row easily.